### PR TITLE
Fix infinite loop in abstemious HTML plugin

### LINKED ISSUE #381: marky-markdown can go into an infinite loop rendering
we found a README in production that forces marky-markdown into an infinite rendering loop in v10:

https://gist.github.com/bcoe/d03e7002657a83ae6a475b492cf10539

we've downgraded to v9, and the file is able to be rendered; any thoughts @revin?

### PR DESCRIPTION
**Pull request by revin**
Found the issue causing an infinite loop in long documents with many consecutive HTML blocks; the rule wasn't doing an early exit when invoked in silent mode (which invocation happens when rules call other rules and only want to know if they match, without mutating the parsing state).

This also drastically speeds up rendering said documents, which is nice. I didn't make any official timing comparisons, but while hunting this down I instrumented the plugin with a silly level of logging output, and the test runs had on the order of 50x fewer lines in the logging because this fixes what ends up being a silly amount of recursion. Bonus!

Fixes #381

### CODE REVIEW

**Comment by revin** [comment]
@bcoe I think this is ready to go. Care to verify?

File: test/markdown.js [hunk 1]
```diff
@@ -356,5 +357,28 @@ describe('markdown processing', function () {
     it('processes a table', function () {
       assert.equal($('table').length, 1)
     })
+
+    it('properly renders long documents containing many consecutive HTML blocks', function (done) {
+      // the failure mode here is an infinite loop that mocha can't interfere
+      // with, so we need to set up an environment that can do it
+      //
+      function handleException (e) {
```
Comment (bcoe) [inline test/markdown.js]: aha! I better understand what the issue was now -- you couldn't even get timeouts?

**Comment by bcoe** [comment]
This is great; I'm going to submit a pull request that adds support for the debugging build step, which we can try out on the next release.

File: test/markdown.js (hunk 1 above)
Comment (revin) [inline test/markdown.js]: Yeah it just sat there running forever. Easily replicated with a `while (1) {}` loop 😕

**Pull request merged**